src/consensus/tx_verify.cpp

```diff
@@ -110,60 +110,21 @@ bool CheckTransaction(const CTransaction& tx, bool fZerocoinActive, CValidationS
     }
 
     std::set<COutPoint> vInOutPoints;
-    std::set<CBigNum> vZerocoinSpendSerials;
-    int nZCSpendCount = 0;
-
     for (const CTxIn& txin : tx.vin) {
         // Check for duplicate inputs
         if (vInOutPoints.count(txin.prevout))
             return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-duplicate");
-
-        //duplicate zcspend serials are checked in CheckZerocoinSpend()
         if (!txin.IsZerocoinSpend()) {
             vInOutPoints.insert(txin.prevout);
-        } else if (!txin.IsZerocoinPublicSpend()) {
-            nZCSpendCount++;
-        }
-    }
-
-    if (fZerocoinActive) {
-        if (nZCSpendCount > consensus.ZC_MaxSpendsPerTx)
-            return state.DoS(100, error("CheckTransaction() : there are more zerocoin spends than are allowed in one transaction"));
-
-        //require that a zerocoinspend only has inputs that are zerocoins
-        if (tx.HasZerocoinSpendInputs()) {
-            for (const CTxIn& in : tx.vin) {
-                if (!in.IsZerocoinSpend() && !in.IsZerocoinPublicSpend())
-                    return state.DoS(100,
-                                     error("CheckTransaction() : zerocoinspend contains inputs that are not zerocoins"));
-            }
-
-            // Do not require signature verification if this is initial sync and a block over 24 hours old
-            bool fVerifySignature = !IsInitialBlockDownload() && (GetTime() - chainActive.Tip()->GetBlockTime() < (60*60*24));
-            if (!CheckZerocoinSpend(tx, fVerifySignature, state, fFakeSerialAttack))
-                return state.DoS(100, error("CheckTransaction() : invalid zerocoin spend"));
```

Comment:
In c15ce7f2893e9a499b7ff6bbb7761d9084b68b59,
This isn't a move-only commit, `CheckZerocoinSpend` function call was removed and not re-added. Need to add it to the `ContextualCheckZerocoinTx` function.

Would be good to have a test validating this changes.

Reply:
> This isn't a move-only commit

This isn't tagged as a move-only commit

Reply:
Also all blocks with a zerocoin spend are over 24 hours old, so no signature verification was already being performed.
The remaining consensus checks are basically the same removed here: 596056cdfa4d34385613d9bfbda3467c874735dc, with the same reasoning.

Reply:
Isn't the same reasoning (my commit message there is not totally accurate). 
Read the comment there, it's **double** checking that the serials were not spent already. Thus why i removed it. The check is still done in `ConnectBlock`, line 1568.

Reply:
> This isn't tagged as a move-only commit

Then better to change the commit message as it is describing a move-only commit:
 "[Refactoring] move zerocoin contextual checks out of CheckTransaction".

Reply:
Before adding this changes, we should add a "rollback to last checkpoint" functionality. So in case of receiving bad blocks when the node is syncing and following a bad path, the node can get back to the right chain rapidly without getting stalled.

Reply:
Added some more refactoring/cleanup